keps/0013-pull-based-audit-events-stream.md

```diff
@@ -0,0 +1,199 @@
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+This proposal aims at providing a pull-based stream for audit logging. This would look similar to a watch request to the API server, and provide stream of audit events. The stream does not return any previous events (i.e. no associated storage), it would only stream events that occurred after the stream was opened. Two non resource-url endpoints are installed to provide the stream. '/audits/' for all audit events generated in the API server, and '/audits/{namespace}' for audits events in the specified namespace. And query parameters could be used to get subset of audit events.
+
+## Motivation
+The current audit backends (log & webhook) require very high privileges (file access to the master) or prior configuration (webhook). There are a number of use cases for dynamic and unprivileged access to the audit logs. Examples include:
+- allow tenants to get audit events in their own namespace
+- provide a subset of audit events, so it will be easier to query and analyze (e.g. https://github.com/kubernetes/kubernetes/issues/56683)
+
+
+### Goals
+- Provide a new pull based audit event stream, which returns events that occurred after the stream was opened.
+- Allow users go get a subset of audit events.
+- Access control. Allow unprivileged users to get audits in their own namespace.
+- Bandwidth limit in server scope or per-namespace.
+
+### Non-Goals
+- Access control at user/group level(e.g. allow user A to only get audits events from users A,B,C)
+
+
+## Proposal
+
+### Make audit backends dynamicly registerable
+
+```golang
+// Registry is a decorator for the audit backend, it allows caller to dynamicly register/unregister new backends to it.
+type Registry interface {
+       Backend
+
+       // Register register a new Backend to the registry.
+       Register(Backend)
+
+       // UnRegister remove the Backend from registry.
+       UnRegister(Backend)
+}
+```
+When a request comes, a new backend is registered to the registry. Audit Events sent to this backend will be sent to users in the end.
+
+### endpoints and query parameters
+
+|endpoint            |definition                                  |supported paramsters                                 |
+|--------------------|--------------------------------------------|-----------------------------------------------------|
+|/audits             |all audit events generated in the API server|username, group, namespace, apiGroup, resource, verbs|
+|/audits/{namespaces}|audits events in the specified namespace    |username, group, apiGroup, resource, verbs           |
+
+definition of query parameters
+```golang
+type queryParm struct {
+       // username in the audit event.
+       // optional
+       username string
+       // groups in the audit event, if all specified groups exist in the audit event, then it will be sent to user.
+       // optional
+       groups []string
```

Comment:
I thought you said username & groups were out of scope under "non goals"?

Reply:
Will remove it.